build-system/tasks/visual-diff/index.js

```diff
@@ -595,6 +598,19 @@ async function ensureOrBuildAmpRuntimeInTestMode_() {
   }
 }
 
+function runYarnForOptionalDependencies_() {
+  try {
+    require.resolve('puppeteer');
+    require.resolve('@percy/puppeteer');
+  } catch (er) {
+    log('Info', 'Running', colors.cyan('yarn'), 'to add optional packages...');
+    execOrDie('yarn install --force');
```

Comment:
From reading the [docs](https://yarnpkg.com/en/docs/cli/install#toc-yarn-install-force), `yarn install --force` will re-fetch _all_ packages, not just the two that are missing. I wonder if there's a way to install just the two missing packages here.

Another undesired side effect of fetching all packages is that the custom patches applied by `gulp update-packages` will be undone, causing future build / test commands to fail.

https://github.com/ampproject/amphtml/blob/28aae733d08a62a06003e1a390452c026c61a955/build-system/tasks/update-packages.js#L194-L197

Maybe you can try running `yarn add --optional puppeteer @percy/puppeteer` and see if the existing version numbers are respected?

Reply:
Unfortunately it doesn't. Currently looking into other ways to avoid `yarn install --force`